pkg/ec2provider/ec2provider.go

```diff
@@ -0,0 +1,274 @@
+	}
+}
+
+func newSession(roleARN string, qps int, burst int) *session.Session {
+	// Initial credentials loaded from SDK's default credential chain, such as
+	// the environment, shared credentials (~/.aws/credentials), or EC2 Instance
+	// Role.
+
+	sess := session.Must(session.NewSession())
+	if aws.StringValue(sess.Config.Region) == "" {
+		ec2metadata := ec2metadata.New(sess)
+		regionFound, err := ec2metadata.Region()
+		if err != nil {
+			logrus.WithError(err).Fatal("Region not found in shared credentials, environment variable, or instance metadata.")
+		}
+		sess.Config.Region = aws.String(regionFound)
+	}
+
+	if roleARN != "" {
+		logrus.WithFields(logrus.Fields{
+			"roleARN": roleARN,
+		}).Infof("Using assumed role for EC2 API")
+
+		rateLimitedClient, err := httputil.NewRateLimitedClient(qps, burst)
+
+		if err != nil {
+			logrus.Errorf("Getting error = %s while creating rate limited client ", err)
+		}
+
+		ap := &stscreds.AssumeRoleProvider{
+			Client:   sts.New(sess, aws.NewConfig().WithHTTPClient(rateLimitedClient).WithSTSRegionalEndpoint(endpoints.RegionalSTSEndpoint)),
+			RoleARN:  roleARN,
+			Duration: time.Duration(60) * time.Minute,
+		}
+
+		sess.Config.Credentials = credentials.NewCredentials(ap)
+	}
+	return sess
+}
+
+func (p *ec2ProviderImpl) setPrivateDNSNameCache(id string, privateDNSName string) {
+	p.privateDNSCache.dnsCacheLock.Lock()
+	defer p.privateDNSCache.dnsCacheLock.Unlock()
+	p.privateDNSCache.dnsCache[id] = privateDNSName
+}
+
+func (p *ec2ProviderImpl) setRequestInQueueForInstanceId(id string) {
+	p.requestQueue.requestQueueLock.Lock()
+	p.requestQueue.requestQueueMap[id] = true
+	p.requestQueue.requestQueueLock.Unlock()
+}
+
+func (p *ec2ProviderImpl) unsetRequestInQueueForInstanceId(id string) {
+	p.requestQueue.requestQueueLock.Lock()
+	delete(p.requestQueue.requestQueueMap, id)
+	p.requestQueue.requestQueueLock.Unlock()
+}
+
+func (p *ec2ProviderImpl) getRequestInQueueForInstanceId(id string) bool {
+	p.requestQueue.requestQueueLock.RLock()
+	_, ok := p.requestQueue.requestQueueMap[id]
+	p.requestQueue.requestQueueLock.RUnlock()
+	return ok
+}
+
+func (p *ec2ProviderImpl) getRequestInQueueSize() int {
+	p.requestQueue.requestQueueLock.RLock()
+	length := len(p.requestQueue.requestQueueMap)
+	p.requestQueue.requestQueueLock.RUnlock()
+	return length
+}
+
+func (p *ec2ProviderImpl) getPrivateDNSNameCache(id string) (string, error) {
+	p.privateDNSCache.dnsCacheLock.RLock()
+	defer p.privateDNSCache.dnsCacheLock.RUnlock()
+	name, ok := p.privateDNSCache.dnsCache[id]
+	if ok {
+		return name, nil
+	}
+	return "", errors.New("instance id not found")
+}
+
+// GetPrivateDNS looks up the private DNS from the EC2 API
+func (p *ec2ProviderImpl) GetPrivateDNSName(id string) (string, error) {
+	privateDNSName, err := p.getPrivateDNSNameCache(id)
+	if err == nil {
+		return privateDNSName, nil
+	}
+	logrus.Infof("Missed the cache for the InstanceId = %s Verifying if its already in requestQueue ", id)
```

Comment:
Can you make this a debug log? this is a little too noisy

Reply:
Sure , added as Debug log